cpp/include/cuvs/neighbors/cagra.hpp

```diff
@@ -557,6 +556,82 @@ struct index : cuvs::neighbors::index {
   raft::device_matrix_view<const IdxT, int64_t, raft::row_major> graph_view_;
   std::unique_ptr<neighbors::dataset<dataset_index_type>> dataset_;
 };
+/**
+ * @}
+ */
+
+/**
+ * @defgroup cagra_cpp_composite_index CAGRA composite index type
+ * @{
+ */
+
+/**
+ * @brief Lightweight composite kNN index for CAGRA.
+ *
+ * This class aggregates logically multiple CAGRA indices into a single composite index,
+ * providing a unified interface for kNN search. It is a lightweight structure
+ * that does not own or manage the lifecycle of the underlying indices; instead,
+ * it holds non-owning pointers to them.
+ *
+ * All sub-indices within the composite index **must share the same distance metric
+ * and dimensionality**.
+ *
+ * @tparam T    Data element type.
+ * @tparam IdxT Index type representing dataset.extent(0), used for vector indices.
+ */
+
+template <typename T, typename IdxT>
+struct composite_index {
+  template <typename Container>
+  explicit composite_index(Container&& indices) : sub_indices(std::forward<Container>(indices))
+  {
+    RAFT_EXPECTS(!sub_indices.empty(), "composite_index requires at least one sub-index.");
+
+    for (auto* idx : sub_indices) {
+      RAFT_EXPECTS(idx != nullptr, "sub_indices contains a null pointer.");
+    }
+
+    auto& first_index = *sub_indices.front();
+    metric_           = first_index.metric();
+    dim_              = first_index.dim();
+    size_             = 0;
+
+    for (auto* idx : sub_indices) {
+      RAFT_EXPECTS(idx->metric() == metric_, "All sub-indices must have the same metric.");
+      RAFT_EXPECTS(idx->dim() == dim_, "All sub-indices must have the same dim.");
+      size_ += idx->size();
+    }
+  }
+
+ public:
+  composite_index(const composite_index& other)            = default;
+  composite_index& operator=(const composite_index& other) = default;
+
+  composite_index(composite_index&& other) noexcept            = default;
+  composite_index& operator=(composite_index&& other) noexcept = default;
+
+  constexpr inline auto metric() const noexcept -> cuvs::distance::DistanceType { return metric_; }
+
+  constexpr inline auto size() const noexcept -> IdxT { return size_; }
+
+  constexpr inline auto dim() const noexcept -> uint32_t { return dim_; }
+
+  constexpr inline auto graph_degree() const noexcept -> uint32_t
+  {
+    return sub_indices.front()->graph_degree();
+  }
+
+  constexpr inline auto num_indices() const noexcept -> uint32_t { return sub_indices.size(); }
+
+ public:
+  std::vector<cuvs::neighbors::cagra::index<T, IdxT>*> sub_indices;
```

Comment:
I know we are planning a potential overhaul of composite index for the next iteration of merging, but in the meantime, it seems like a fairly straightforward change to introduce a template type to capture the `neighbors::xxx::index` type for the `composite_index`. Can we do that, and move this to `common.hpp`? That would be a good start anyways. Then the `cagra::composite_index` can extend `neighbors::composite_index` and provide the additional accessor methods for the cagra-specific things.

Reply:
I know we are pushing this to a follow-on, but just want to make sure we're clear about the expectations for that design change.